datalad_registry/search.py

```diff
@@ -0,0 +1,279 @@
+
+// Let's make it into a rule so we could have transformer for it
+// and unescape escaped \s
+?quoted_string: ESCAPED_STRING // /"((?:\.|[^\"])*)"/
+WORD: /[-_.*?+\w]+/
+
+// :operations to aim for
+!op: "?"   // string present somewhere within
+    |"="   // equal to the value as a string
+    |">"   // greater to the value as a string
+    |"~"   // regex search, should be able to use ^ and $ to anchor if desired
+// TODO:
+//  - come up with transformers, e.g. (int) or `!int` to precede :? etc which
+//    should change how operations like =, >
+//  - operator "in" might come handy
+//  - also to make it case-sensitive since to make most useful - by default are\
+//    case insensitive!
+
+%import common.ESCAPED_STRING
+%import common.WS
+%ignore WS
+"""
+
+
+def _dump_grammar():
+    # Could be used for debugging on the web -- the schema with those fields embedded
+    with open("/tmp/grammar.lark", "w") as f:
+        f.write(grammar)
+
+
+# _dump_grammar()
+
+
+class GrammarValueError(ValueError, GrammarError):
+    """Indicates that an unknown field was specified in the search query.
+
+    Making a dedicated exception subclassing GrammarError since that is
+    the class which lark re-raises, and otherwise would wrap into VisitError
+    """
+
+    pass
+
+
+@v_args(inline=True)  # Affects the signatures of the methods
+class SearchQueryTransformer(Transformer):
+    """Convert the parsed search query into SQLAlchemy expressions."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Additional initialization here if needed
+
+    def or_search(self, *args) -> ColumnElement[bool]:
+        # args will be a list of the arguments to the OR operation
+        lgr.debug("OR search: %s", args)
+        return or_(*args)
+
+    def and_search(self, *args) -> ColumnElement[bool]:
+        # args will be a list of the arguments to the AND operation
+        lgr.debug("AND search: %s", args)
+        return and_(*args)
+
+    def not_expr(self, arg) -> ColumnElement[bool]:
+        # args will be a single-element list containing the NOT argument
+        lgr.debug("AND search: %s", arg)
+        return not_(arg)
+
+    def get_field_select_search(
+        self, metadata_field_l, metadata_extractors_l, value
+    ) -> CollectionAggregate[bool]:
+        assert metadata_field_l.data.value == "metadata_field"
+        if isinstance(metadata_extractors_l, Token):
+            extractors = [metadata_extractors_l.value]
+        else:  # was more than one
+            assert isinstance(metadata_extractors_l, Tree)
+            assert metadata_extractors_l.data.value == "metadata_extractors"  # type: ignore
+            extractors = list(
+                map(self._get_str_value, metadata_extractors_l.children)  # type: ignore
+            )
+        if not extractors:
+            raise GrammarValueError(
+                f"No extractors were specified in {metadata_extractors_l}"
+            )
```

Comment:
I think they are not even reachable per se since lark grammar wouldn't allow for such a case. So it is more of an AssertionError here in particular. But I take it as a defensive coding to code under assumption we might get there. I would have just added the skip for coverage here.